plugins/modules/cloud/misc/proxmox_disk.py

```diff
@@ -0,0 +1,735 @@
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        required_together=[('api_token_id', 'api_token_secret')],
+        required_one_of=[('name', 'vmid'), ('api_password', 'api_token_id')],
+        required_if=[
+            ('state', 'present', ['storage']),
+            ('state', 'resized', ['size'])
+        ],
+        required_by={
+            'target_disk': 'target_vmid',
+            'mbps_max': 'mbps',
+            'mbps_rd_max': 'mbps_rd',
+            'mbps_wr_max': 'mbps_wr',
+            'bps_max_length': 'mbps_max',
+            'bps_rd_max_length': 'mbps_rd_max',
+            'bps_wr_max_length': 'mbps_wr_max',
+            'iops_max': 'iops',
+            'iops_rd_max': 'iops_rd',
+            'iops_wr_max': 'iops_wr',
+            'iops_max_length': 'iops_max',
+            'iops_rd_max_length': 'iops_rd_max',
+            'iops_wr_max_length': 'iops_wr_max'
+        },
+        supports_check_mode=False,
+        mutually_exclusive=[
+            ('target_vmid', 'target_storage'),
+            ('mbps', 'mbps_rd'),
+            ('mbps', 'mbps_wr'),
+            ('iops', 'iops_rd'),
+            ('iops', 'iops_wr'),
+            ('import_from', 'size'),
+        ]
+    )
+
+    proxmox = ProxmoxDiskAnsible(module)
+
+    disk = module.params['disk']
+    # Verify disk name has appropriate name
+    disk_regex = compile(r'^([a-z]+)([0-9]+)$')
+    disk_bus = sub(disk_regex, r'\1', disk)
+    disk_number = int(sub(disk_regex, r'\2', disk))
+    if disk_bus not in proxmox.supported_bus_num_ranges:
+        proxmox.module.fail_json(msg='Unsupported disk bus: %s' % disk_bus)
+    elif disk_number not in proxmox.supported_bus_num_ranges[disk_bus]:
+        bus_range = proxmox.supported_bus_num_ranges[disk_bus]
+        proxmox.module.fail_json(msg='Disk %s number not in range %s..%s ' % (disk, bus_range[0], bus_range[-1]))
+
+    name = module.params['name']
+    state = module.params['state']
+    vmid = module.params['vmid'] or proxmox.get_vmid(name)
+
+    # Ensure VM id exists and retrieve its config
+    vm = None
+    vm_config = None
+    try:
+        vm = proxmox.get_vm(vmid)
+        vm_config = proxmox.proxmox_api.nodes(vm['node']).qemu(vmid).config.get()
+    except Exception as e:
+        proxmox.module.fail_json(msg='Getting information for VM %s failed with exception: %s' % (vmid, str(e)))
+
+    if state == 'present':
+        try:
+            if proxmox.create_disk(disk, vmid, vm, vm_config):
+                module.exit_json(changed=True, vmid=vmid, msg="Disk %s created in VM %s" % (disk, vmid))
+            else:
+                module.exit_json(changed=False, vmid=vmid, msg="Disk %s already exists in VM %s." % (disk, vmid))
+        except Exception as e:
+            module.fail_json(vmid=vmid, msg='Unable to create disk %s in VM %s: %s' % (disk, vmid, str(e)))
+
+    # Do not try to perform actions on missing disk
+    if disk not in vm_config and state in ['updated', 'resized', 'moved']:
```

Comment:
I'm wondering whether it makes sense to have a state which allows to combine `present` and `updated`. Usually `state=present` does exactly that, but especially for resources like disks it's always good if you can disable the "update if already exists" part, so having two separate states (or a boolean flag) is IMO a good idea here.

Reply:
I'm sorry, I don't get your point. In current implementation:

- `state=present` can only create new disk or rewrite existing with `force` (it does not change options);
- `state=update` can change options only for existing disk (it does not create new).

Reply:
@castorsky Felix point is doing the same thing, only using a different language. Ansible is meant to be idempotent and  a declarative idiom, based on "desired-state", is better suited to achieve that than an imperative idiom.

In his suggestion, `state=present` would create a disk if it does not exist or modify it if it does, so when the user says `state=present` they will know that one such disk will exist after the task is run (unless errors, of course), regardless of the previous state.

Reply:
Now it is clear. :) I will add `update` flag to `state=present` for user to declare his will to change options.

Reply:
Well, a bit of a nit picking here, but I think it would perhaps look more usual to have `state=present` and a `create` flag, with default `true`, so that, this create/update behaviour is the default. If the user *does NOT want* to create a disk if it is not there (i.e., only to update an existing one), then they set `create=false`.

And, in any way you choose to implement it, don't forget to remove the `updated` state. ;-)

Reply:
Okay, I got the point. How do you think will be more elegant to have two boolean flags (`create` and `force`) or one tri-state `create = ['disable', 'regular', 'forced']`?